python/tests/test_coalrate.py

```diff
@@ -0,0 +1,357 @@
+# MIT License
+#
+# Copyright (c) 2024 Tskit Developers
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+"""
+Test cases for coalescence time distribution objects in tskit.
+"""
+import itertools
+
+import msprime
+import numpy as np
+import pytest
+
+import tests
+import tskit
+
+
+def naive_coalescing_pairs(ts, sample_set_0, sample_set_1):
+    """
+    Count pairwise coalescences tree by tree, by enumerating nodes in each
+    tree. For a binary node, the number of pairs of samples that coalesce in a
+    given node is the product of the number of samples subtended by the left
+    and right child. For higher arities, the count is summed over all possible
+    pairs of children.
+    """
+    output = np.zeros(ts.num_nodes)
+    for t in ts.trees():
+        sample_counts = np.zeros((ts.num_nodes, 2), dtype=np.int32)
+        for n in t.postorder():
+            samples = list(t.samples(n))
+            sample_counts[n, 0] = np.intersect1d(samples, sample_set_0).size
+            sample_counts[n, 1] = np.intersect1d(samples, sample_set_1).size
```

Comment:
Could do this in a single postorder pass I think? Just for clarity of what the algorithm is doing.